### PR TITLE
[RHPAM-2435] PIM - Operator Integration

### PR DESCRIPTION
**Pull request by xiezhang7**
Deploy Process Instance Migration image on Openshift using the kie-cloud-operator.
https://issues.redhat.com/browse/RHPAM-2435

### CODE REVIEW

**Comment by xiezhang7** [comment]
> @xiezhang7 there are two many changes for just PIM Operator integration. I see that in commits there are also updates for Operator SDK updates. Can you please remove these updates from this PR, there is already one existing PR for operator sdk update opened #309

@jakubschwan I'll rebase this PR once the master branch accepts 7.7.0 changes and the operator-sdk updates PR #309 is merged. Then this PR will only have changes related to PIM integration. Thanks.

**Comment by tchughesiv** [comment]
please rebase and update to use 7.7.0 configs

**Comment by xiezhang7** [comment]
This pull request is ready for review. Thanks.

File: pkg/apis/app/v2/kieapp_types.go [hunk 1]
```diff
@@ -361,6 +365,7 @@ type EnvTemplate struct {
 	Servers       []ServerTemplate    `json:"servers,omitempty"`
 	SmartRouter   SmartRouterTemplate `json:"smartRouter,omitempty"`
 	Auth          AuthTemplate        `json:"auth,omitempty"`
+	PIM           PIMTemplate         `json:"pim,omitempty"`
```
Comment (ruromero) [inline pkg/apis/app/v2/kieapp_types.go]: I would capitalize only the first letter.

File: pkg/apis/app/v2/kieapp_types.go [hunk 2]
```diff
@@ -478,6 +483,27 @@ type RoleMapperTemplate struct {
...
 }
 
+// PIMObject
+type PIMObject struct {
+	KieAppObject `json:",inline"`
+	Database     DatabaseObject `json:"database,omitempty"`
+}
+
+// PIMTemplate
+type PIMTemplate struct {
```
Comment (ruromero) [inline pkg/apis/app/v2/kieapp_types.go]: I prefer to capitalize only the first letter

File: pkg/apis/app/v2/kieapp_types.go [hunk 3]
```diff
@@ -478,6 +483,27 @@ type RoleMapperTemplate struct {
 	RoleMapperAuthConfig `json:",inline"`
 }
 
+// PIMObject
+type PIMObject struct {
```
Comment (ruromero) [inline pkg/apis/app/v2/kieapp_types.go]: I prefer to capitalize only the first letter

File: pkg/apis/app/v2/kieapp_types.go [hunk 4]
```diff
@@ -478,6 +483,27 @@ type RoleMapperTemplate struct {
...
+	KieAppObject `json:",inline"`
```
Comment (ruromero) [inline pkg/apis/app/v2/kieapp_types.go]: Can't have this field because it doesn't have a keystore object.

File: pkg/apis/app/v2/kieapp_types.go [hunk 5]
```diff
@@ -478,6 +483,27 @@ type RoleMapperTemplate struct {
...
+// PIMTemplate
+type PIMTemplate struct {
+	Image      string         `json:"image,omitempty"`
+	ImageTag   string         `json:"imageTag,omitempty"`
+	KieServers []KieServer    `json:"kieServers,omitempty"`
+	Database   DatabaseObject `json:"database,omitempty"`
+}
+
+// KieServer
+type KieServer struct {
```
Comment (ruromero) [inline pkg/apis/app/v2/kieapp_types.go]: To prevent confusion. I suggest to rename it to `KieServerClient`

File: pkg/controller/kieapp/defaults/defaults.go [hunk 6]
```diff
@@ -817,3 +822,91 @@ func setDefaults(cr *api.KieApp) {
...
+	} else {
+		pimEnv.PIM.Omit = true
+	}
+
+	return env, nil
+}
+
+func mergePIMDB(service api.PlatformService, cr *api.KieApp, env api.Environment, envTemplate api.EnvTemplate) (api.Environment, error) {
+	if envTemplate.PIM.Database.Type == api.DatabaseExternal {
+		return api.Environment{}, fmt.Errorf("external database type is not supported for Process Instance Migration")
```
Comment (ruromero) [inline pkg/controller/kieapp/defaults/defaults.go]: If it is not supported then remove it from the CRD

File: deploy/crds/kieapp.crd.yaml [hunk 7]
```diff
@@ -542,6 +542,127 @@ spec:
...
+                          required:
+                            - type
+                          properties:
+                            type:
+                              type: string
+                              description: Database type to use
+                              enum:
+                                - mysql
+                                - postgresql
+                                - external
```
Comment (ruromero) [inline deploy/crds/kieapp.crd.yaml]: Remove as it is not yet supported

File: deploy/crds/kieapp.crd.yaml [hunk 8]
```diff
@@ -542,6 +542,127 @@ spec:
...
+                                - h2
+                            size:
+                              type: string
+                              description: Size of the PersistentVolumeClaim to create. For example, 100Gi
+                            externalConfig:
```
Comment (ruromero) [inline deploy/crds/kieapp.crd.yaml]: As it is not supported, all the externalConfig should be removed

File: deploy/ui/form.json [hunk 9]
```diff
@@ -1328,8 +1328,52 @@
...
+                }
+              ]
+            }
+          ]
         }
       ]
     }
   ]
-}
+}
```
Comment (ruromero) [inline deploy/ui/form.json]: Add new-line

File: config/7.7.0/pim/pim.yaml [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        template:
+          metadata:
+            name: "[[.ApplicationName]]-process-migration"
+            labels:
+              app: "[[.ApplicationName]]"
+              application: "[[.ApplicationName]]"
+              deploymentConfig: "[[.ApplicationName]]-process-migration"
+              service: "[[.ApplicationName]]-process-migration"
+          spec:
+            serviceAccountName: "[[$.ApplicationName]]-[[$.Constants.Product]]svc"
```
Comment (ruromero) [inline config/7.7.0/pim/pim.yaml]: It doesn't need a serviceAccount.

Comment (xiezhang7) [reply config/7.7.0/pim/pim.yaml]: The `serviceAccount` is required by the `ose-cli` init-container to execute `oc get dc`.

**Comment by xiezhang7** [comment]
Thanks @ruromero . I have made the changes you suggested except for `serviceAccount`. Please let me know if the changes to the CRD is good, because we also need to update the console UI for the CRD changes. Thanks.

**Comment by ruromero** [comment]
@xiezhang7 please, for the next time try to not force-push unless it is a rebase or until the review is finished. It will help me with reviewing the incremental changes.
Thanks

File: deploy/ui/form.json [hunk 11]
```diff
@@ -1328,6 +1328,50 @@
...
+            },
+            {
+              "label": "Image tag",
+              "required": false,
+              "jsonPath": "$.spec.objects.pim.imageTag",
+              "description": "The image tag to use for Process Instance Migration.",
+              "type": "text"
+            },
+            {
+              "label": "Database type",
```
Comment (ruromero) [inline deploy/ui/form.json]: The database size is missing

**Comment by xiezhang7** [comment]
> @xiezhang7 please, for the next time try to not force-push unless it is a rebase or until the review is finished. It will help me with reviewing the incremental changes.
> Thanks

@ruromero sure, I'll avoid doing that. Actually, the changes related to your review comments are in the second commit. Please take a look at that.

File: pkg/apis/app/v2/kieapp_types.go [hunk 12]
```diff
@@ -478,6 +483,34 @@ type RoleMapperTemplate struct {
 	RoleMapperAuthConfig `json:",inline"`
 }
 
+// ProcessMigrationObject
+type ProcessMigrationObject struct {
+	Image    string                         `json:"image,omitempty"`
+	ImageTag string                         `json:"imageTag,omitempty"`
+	Database ProcessMigrationDatabaseObject `json:"database,omitempty"`
```
Comment (ruromero) [inline pkg/apis/app/v2/kieapp_types.go]: You can reuse the `DatabaseObject` although up to the CRD `external` is not supported. That would help the refactoring in the next iteration

File: pkg/apis/app/v2/kieapp_types.go [hunk 13]
```diff
@@ -478,6 +483,34 @@ type RoleMapperTemplate struct {
...
+}
+
+// ProcessMigrationTemplate
+type ProcessMigrationTemplate struct {
+	Image            string                         `json:"image,omitempty"`
+	ImageTag         string                         `json:"imageTag,omitempty"`
+	KieServerClients []KieServerClient              `json:"kieServerClients,omitempty"`
+	Database         ProcessMigrationDatabaseObject `json:"database,omitempty"`
```
Comment (ruromero) [inline pkg/apis/app/v2/kieapp_types.go]: Same. `DatabaseObject` can be used

File: pkg/apis/app/v2/kieapp_types.go [hunk 14]
```diff
@@ -478,6 +483,34 @@ type RoleMapperTemplate struct {
...
+
+// KieServerClient
+type KieServerClient struct {
+	Host     string `json:"host,omitempty"`
+	Username string `json:"username,omitempty"`
+	Password string `json:"password,omitempty"`
+}
+
+// ProcessMigrationDatabaseObject
+type ProcessMigrationDatabaseObject struct {
```
Comment (ruromero) [inline pkg/apis/app/v2/kieapp_types.go]: Not needed IMO

File: deploy/ui/form.json (hunk 11 above)
Comment (xiezhang7) [reply deploy/ui/form.json]: Once changes to console UI is finished, I'll update this pull request to include these changes.

File: pkg/apis/app/v2/kieapp_types.go (hunk 12 above)
Comment (xiezhang7) [reply pkg/apis/app/v2/kieapp_types.go]: The `crd_validation_test` will fail for "Discrepancy between CRD and Struct". Is there a way we can skip the test for `external`?

Comment (ruromero) [reply pkg/apis/app/v2/kieapp_types.go]: Right, I didn't think about the crd_validation_test.

**Comment by ruromero** [comment]
As discussed, let's add back the external db support and fully implement it given that it has been pushed to the next sprint

**Comment by tchughesiv** [comment]
please rebase

**Comment by xiezhang7** [comment]
Need to rebase the code, and add back the [PIM images](https://github.com/kiegroup/kie-cloud-operator/pull/396#issuecomment-603828353) for #396 .

**Comment by spolti** [comment]
@xiezhang7 can you please rebase?

**Comment by xiezhang7** [comment]
> @xiezhang7 can you please rebase?

@spolti I am working on adding the support for external DB. I'll rebase and update this pull request once that's ready. Thanks.

**Comment by xiezhang7** [comment]
@ruromero @bmozaffa @jakubschwan this PR is ready for review. I'll work with @swati-kale to update the `form.json` in this PR, once the CR changes are approved. Thanks.

File: config/7.8.0/pim/dbs/postgresql.yaml [hunk 15]
```diff
@@ -0,0 +1,154 @@
+## KIE ProcessMigration BEGIN
+processMigration:
+  ## KIE ProcessMigration deployment config BEGIN
+  deploymentConfigs:
```
Comment (ruromero) [inline config/7.8.0/pim/dbs/postgresql.yaml:4]: I wonder whether we can somehow reuse the existing configurations for kieservers. That introduces a lot of duplicated configuration.

File: pkg/apis/app/v2/kieapp_types.go [hunk 16]
```diff
@@ -348,21 +352,28 @@ const (
...
-	ExternalConfig   *ExternalDatabaseObject `json:"externalConfig,omitempty"`
+	InternalDatabaseObject `json:",inline"`
+	ExternalConfig         *ExternalDatabaseObject `json:"externalConfig,omitempty"`
 }
 
-// ExternalDatabaseObject configuration definition of an external database
-type ExternalDatabaseObject struct {
+// ProcessMigrationDatabaseObject Defines how a Process Migration server will manage
+// and create a new Database or connect to an existing one
+type ProcessMigrationDatabaseObject struct {
```
[the remark this answers is not in the document]
Comment (xiezhang7) [reply pkg/apis/app/v2/kieapp_types.go:361]: Because the `ExternalConfig` of `DatabaseObject` has `Dialect`, `Name`, `Host`, `Port`, `NonXA`. I don't think these are applicable to the data source configuration for process migration.

[the remark this answers is not in the document]
Comment (ruromero) [reply pkg/apis/app/v2/kieapp_types.go:361]: After reading the documentation and scripts in KIE Server it seems these properties won't be needed because they are used in mysql and postgres but it is possible to configure the datasources without them.
In EAP they are set as datasource properties so in Thorntail it should be set like this:
`thorntail.datasources.xa-data-sources.KEY.xa-datasource-properties.ServerName.value = Host`
`thorntail.datasources.xa-data-sources.KEY.xa-datasource-properties.PortNumber.value = Port`
The dialect is used in the kie server scripts to decide what to require in terms of properties depending on the value so no need to use it here either.

File: config/7.8.0/pim/dbs/postgresql.yaml (hunk 15 above)
Comment (xiezhang7) [reply config/7.8.0/pim/dbs/postgresql.yaml]: @ruromero I have moved the deployment of mysql and postgresql to separate templates and reused by both kie server and process migration. Please take a look. Thanks.

File: deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml [hunk 17]
```diff
@@ -482,6 +492,8 @@ spec:
     name: rhpam-businesscentral-monitoring-rhel8
   - image: registry.redhat.io/rhpam-7/rhpam-smartrouter-rhel8@sha256:34ccf7889abce588bf0db5956b1747d785934dc5630135094949318bef49a70e
     name: rhpam-smartrouter-rhel8
+  - image: registry.redhat.io/rhpam-7/rhpam-process-migration-rhel8:7.7.0
```
Comment (tchughesiv) [inline deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml:495]: does this image exist in 7.7.0? and will one exist for 7.7.1? If not, we need to only add for 7.8.0 releases or later.

File: deploy/examples/v2/process_migration_externaldb.yaml [hunk 18]
```diff
@@ -0,0 +1,22 @@
+apiVersion: app.kiegroup.org/v2
+kind: KieApp
+metadata:
+  name: process-migration
+spec:
+  environment: rhpam-trial
+  objects:
+    processMigration:
+      image: "rhpam-process-migration-rhel8-mssql"
```
Comment (tchughesiv) [inline deploy/examples/v2/process_migration_externaldb.yaml:9]: what is this image? is it something RH will provide?

File: pkg/apis/app/v2/kieapp_types.go [hunk 19]
```diff
@@ -199,10 +201,12 @@ type KieAppObject struct {
...
 
 type Environment struct {
-	Console     CustomObject   `json:"console,omitempty"`
-	SmartRouter CustomObject   `json:"smartRouter,omitempty"`
-	Servers     []CustomObject `json:"servers,omitempty"`
-	Others      []CustomObject `json:"others,omitempty"`
+	Console          CustomObject   `json:"console,omitempty"`
+	SmartRouter      CustomObject   `json:"smartRouter,omitempty"`
+	Servers          []CustomObject `json:"servers,omitempty"`
+	ProcessMigration CustomObject   `json:"processMigration,omitempty"`
```
Comment (tchughesiv) [inline pkg/apis/app/v2/kieapp_types.go:207]: I assume this means there can only be one PIM object per deployment?

File: config/7.8.0/pim/process-migration.yaml [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+        triggers:
+          - type: ImageChange
+            imageChangeParams:
+              automatic: true
+              containerNames:
+                - "[[.ApplicationName]]-process-migration"
+              from:
+                kind: ImageStreamTag
+                namespace: openshift
+                name: "[[.ProcessMigration.Image]]:[[.ProcessMigration.ImageTag]]"
```
Comment (tchughesiv) [inline config/7.8.0/pim/process-migration.yaml:25]: image triggers with imagestreams are the old way of deploying. since we've moved to image sha's and disconnected installs, we default to direct image url pulls instead using -
https://github.com/kiegroup/kie-cloud-operator/blob/b675341b04e1d0cac9553f001be32e528731413d/config/7.8.0/pim/process-migration.yaml#L40

reference other object yaml like console and smart router to determine how this is done.
https://github.com/kiegroup/kie-cloud-operator/blob/b675341b04e1d0cac9553f001be32e528731413d/pkg/controller/kieapp/defaults/defaults.go#L261-L279

you'll probably want to add an ImageURL field to PIM template object.
https://github.com/kiegroup/kie-cloud-operator/blob/b675341b04e1d0cac9553f001be32e528731413d/config/7.8.0/common.yaml#L13-L41

File: deploy/examples/v2/process_migration_externaldb.yaml (hunk 18 above)
Comment (xiezhang7) [reply deploy/examples/v2/process_migration_externaldb.yaml:9]: This is an example image. It is supposed to be based on the process migration image and packaged with the jdbc driver for the external db connection. Please let me know if there is a  way to make it less ambiguous. Thanks.

File: pkg/apis/app/v2/kieapp_types.go (hunk 19 above)
Comment (xiezhang7) [reply pkg/apis/app/v2/kieapp_types.go:207]: Yes, only one PIM object per deployment.

File: deploy/examples/v2/process_migration_externaldb.yaml (hunk 18 above)
Comment (tchughesiv) [reply deploy/examples/v2/process_migration_externaldb.yaml:9]: maybe add a commented out description above it?

File: deploy/examples/v2/process_migration.yaml [hunk 21]
```diff
@@ -0,0 +1,13 @@
+apiVersion: app.kiegroup.org/v2
+kind: KieApp
+metadata:
+  name: process-migration
+spec:
+  environment: rhpam-trial
+  objects:
+    processMigration:
+      image: "rhpam-process-migration-rhel8"
```
Comment (tchughesiv) [inline deploy/examples/v2/process_migration.yaml:9]: why is this here? isn't this the default image?

Comment (xiezhang7) [reply deploy/examples/v2/process_migration.yaml]: Removed

File: deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml (hunk 17 above)
Comment (xiezhang7) [reply deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml]: Removed 7.7.0 and 7.7.1

File: config/7.8.0/pim/process-migration.yaml (hunk 20 above)
Comment (xiezhang7) [reply config/7.8.0/pim/process-migration.yaml:26]: Updated to use the new way of deploying

File: deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml (hunk 17 above)
Comment (ruromero) [reply deploy/catalog_resources/community/7.8/kiecloud-operator.7.8.0.clusterserviceversion.yaml]: The image creation request is for 7.8.0

**Comment by ruromero** [comment]
Waiting for the ConsoleUI changes, otherwise LGTM

**Comment by tchughesiv** [comment]
> Waiting for the ConsoleUI changes, otherwise LGTM

@ruromero are we still waiting for any changes before we can merge this?

File: deploy/ui/form.json (hunk 11 above)
Comment (ruromero) [reply deploy/ui/form.json:1641]: Fields are still missing here

File: deploy/ui/form.json [hunk 22]
```diff
@@ -1619,6 +1619,50 @@
...
+              "type": "dropDown",
+              "required": false,
+              "jsonPath": "$.spec.objects.pim.database.type",
+              "originalJsonPath": "$.spec.objects.pim.database.type",
+              "options": [
```
[the remark this answers is not in the document]
Comment (xiezhang7) [reply deploy/ui/form.json:1646]: Should I remove the changes to `form.json` in this pull request? It seems update to `form.json` is in #420 .

[the remark this answers is not in the document]
Comment (tchughesiv) [reply deploy/ui/form.json:1646]: @swati-kale  ... you and @xiezhang7 need to decide if the form changes warrant a separate PR. Either way, please sync on the changes required.

[the remark this answers is not in the document]
Comment (xiezhang7) [reply deploy/ui/form.json]: Changes to `form.json` reverted.

[the remark this answers is not in the document]
Comment (swati-kale) [reply deploy/ui/form.json]: I did not see this earlier, @xiezhang7 , this  does not look correct as jsonPath has "$..pim...", instead "..processMigration..". 
I have the corresponding PR#420 for the changes you have asked. Do you want me to close that?

[the remark this answers is not in the document]
Comment (xiezhang7) [reply deploy/ui/form.json]: @swati-kale I have reverted the changes in this RP, you can keep RP #420 .

**Comment by xiezhang7** [comment]
> @xiezhang7 can you please ensure you've run `make test` with go 1.12 ? I'm seeing vendor changes that normally occur with go 1.13.
> 
> Afterwards... please commit the changes.

@tchughesiv Reverted the changes to `vendor` and tested with go 1.12. Thanks.

**Comment by tchughesiv** [comment]
thanks @xiezhang7 ... please rebase/squash to a single commit and i think we're ready to merge.

**Comment by xiezhang7** [comment]
> thanks @xiezhang7 ... please rebase/squash to a single commit and i think we're ready to merge.

@tchughesiv commits squashed.

**Comment by bmozaffa** [comment]
🎉🎉🎉
